src/nvim/option.c

```diff
@@ -6905,15 +7040,309 @@ int check_ff_value(char_u *p)
+      break;
+    }
+  }
+  if (t > tabcount) {
+    tab_size = (int)vts[tabcount];
+  }
+
+  return tab_size;
+}
+
+// Find the column on which a tab starts.
+colnr_T tabstop_start(colnr_T col, long ts, long *vts)
+{
+  int tabcount;
+  colnr_T tabcol = 0;
+  int t;
+  int excess;
+
+  if (vts == NULL || vts[0] == 0) {
+    return (col / (int)ts) * (int)ts;
+  }
+
+  tabcount = (int)vts[0];
+  for (t = 1; t <= tabcount; t++) {
+    tabcol += vts[t];
+    if (tabcol > col) {
+      return tabcol - (int)vts[t];
+    }
+  }
+
+  excess = tabcol % vts[tabcount];
+  return excess + ((col - excess) / (int)vts[tabcount]) * (int)vts[tabcount];
+}
+
+// Find the number of tabs and spaces necessary to get from one column
+// to another.
+void tabstop_fromto(colnr_T start_col,
+                    colnr_T end_col,
+                    long ts_arg,
+                    long *vts,
+                    int *ntabs,
+                    int *nspcs)
+{
+  int spaces = end_col - start_col;
+  colnr_T tabcol = 0;
+  int padding = 0;
+  int tabcount;
+  int t;
+  long ts = ts_arg == 0 ? curbuf->b_p_ts : ts_arg;
+
+  if (vts == NULL || vts[0] == 0) {
+    int tabs = 0;
+    int initspc = 0;
+
+    initspc = (int)ts - (start_col % (int)ts);
+    if (spaces >= initspc) {
+      spaces -= initspc;
+      tabs++;
+    }
+    tabs += spaces / ts;
+    spaces -= (spaces / ts) * ts;
+
+    *ntabs = tabs;
+    *nspcs = spaces;
+    return;
+  }
+
+  // Find the padding needed to reach the next tabstop.
+  tabcount = (int)vts[0];
+  for (t = 1; t <= tabcount; t++) {
+    tabcol += vts[t];
+    if (tabcol > start_col) {
+      padding = (int)(tabcol - start_col);
+      break;
+    }
+  }
+  if (t > tabcount) {
+    padding =
+      (int)vts[tabcount] - (int)((start_col - tabcol) % (int)vts[tabcount]);
+  }
+
+  // If the space needed is less than the padding no tabs can be used.
+  if (spaces < padding) {
+    *ntabs = 0;
+    *nspcs = spaces;
+    return;
+  }
+
+  *ntabs = 1;
+  spaces -= padding;
+
+  // At least one tab has been used. See if any more will fit.
+  while (spaces != 0 && ++t <= tabcount) {
+    padding = (int)vts[t];
+    if (spaces < padding) {
+      *nspcs = spaces;
+      return;
+    }
+    *ntabs += 1;
+    spaces -= padding;
+  }
+
+  *ntabs += spaces / (int)vts[tabcount];
+  *nspcs = spaces % (int)vts[tabcount];
+}
+
+// See if two tabstop arrays contain the same values.
+bool tabstop_eq(long *ts1, long *ts2)
+{
+  int t;
+
+  if ((ts1 == 0 && ts2) || (ts1 && ts2 == 0)) {
+    return false;
+  }
+  if (ts1 == ts2) {
+    return true;
+  }
+  if (ts1[0] != ts2[0]) {
+    return false;
+  }
+
+  for (t = 1; t <= ts1[0]; t++) {
+    if (ts1[t] != ts2[t]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Copy a tabstop array, allocating space for the new array.
+int *tabstop_copy(long *oldts)
+{
+  long *newts;
+  int t;
+
+  if (oldts == 0) {
+    return 0;
+  }
+
+  newts = (long *)xmalloc((unsigned)(oldts[0] + 1) * sizeof(long));
```

Comment:
Does CI fail if `xmalloc` is not casted?

Reply:
I'll try.